### ISSUE TITLE
Question about the GaussLegendre basis

### ISSUE THREAD

**Issue by rhebergens** [mfem/mfem#327]
Hello,

We're trying to understand the implementation of the GaussLegendre basis, and gradient, used on triangles when using L2_FECollection, in particular, in
`void L2_TriangleElement::CalcShape`
and
`void L2_TriangleElement::CalcDShape`
both in fe.cpp. Do you maybe have a reference to literature where this is explained in a bit more detail?
For example, when computing the gradient of a P1 function we're expecting a matrix with 1's and -1's, but instead we're getting other values.

We're also wondering what the difference is between using `CalcLegendre(p, x, u);` (which is commented) and `CalcChebyshev(p, x, u);` in `static void CalcBasis` in fe.hpp.

Thanks,
Sander

**Comment by tzanio** [comment]
Hi Sander,

In 1D, the GaussLegendre basis is just an interpolatory basis in the Gaussian quadrature points. This is an example of an *open* basis in our terminology, where the endpoints are not included as degrees of freedom (which explains the slopes you are seeing). 

![gleg](https://user-images.githubusercontent.com/1278247/32060703-70391600-ba24-11e7-8548-ad0afd2f869b.png)

An examples of a *closed* basis is GaussLobatto, which is an interpolatory basis in the Gauss-Lobatto quadrature points, which do include the endpoints of the interval and thus are appropriate for H1-type spaces.

![glob](https://user-images.githubusercontent.com/1278247/32061094-9229b20a-ba25-11e7-8e71-f906c91af219.png)

Another example of a closed basis is the Bernstein positive basis, which is interpolatory only at the endpoints (the rest of the degrees of freedom don't correspond to function values).

![ber](https://user-images.githubusercontent.com/1278247/32060704-70540b18-ba24-11e7-8957-de155296ebd3.png)

For quads and hexes, the GaussLegendre and GaussLobatto bases are just Cartesian products of the 1D ones. Things are more involved for triangles and tets, but basically you start with the 1D dofs on the edges and extend in a triangular lattice on the faces and in the interior. 

@v-dobrev can probably provide more details for this particular case if you are interested. He or @pgmaginot may also have more information about your last question.

Hope this helps,
Tzanio

**Comment by rhebergens** [comment]
Hello Tzanio,
Thanks for the explanation. The reason we were looking at this was because we were getting weird solutions from our code. We started to compare our MFEM matrices to those of another code we have and the matrices were different, so we figured this might be because of the different basis functions that were being used. We eventually found that we had to increase the order of the quadrature used in VectorDomainLFIntegrator and now everything works. Tamas posted this as a new question.
Sander
